rpc/src/responses/json_array.rs

```diff
@@ -1,47 +1,95 @@
 use super::ParseError;
 use serde::{de, Deserialize, Serialize};
-use serde_json::{self, json, Value};
+use serde_json::{self, Value};
+use std::array::IntoIter;
+use std::fmt;
 
 #[derive(Serialize, Deserialize)]
 pub struct JsonArray<T> {
-    flattened_vec: Vec<T>,
+    items: Vec<T>,
 }
 
 impl<T: de::DeserializeOwned> JsonArray<T> {
     pub fn from_response_str(str_to_parse: &str) -> Result<Self, ParseError> {
-        let mut parse_response = serde_json::from_str::<Value>(str_to_parse)?;
-        let parsed_json_array = parse_response
-            .as_array_mut()
-            .ok_or_else(|| generate_none_error("cannot parse initial json string as array"))?;
-
-        let mut flattened_vec = Vec::new();
-        for nested_entry in parsed_json_array {
-            let parsed_item_value = match nested_entry.is_array() {
-                true => unwrap_item_in_nested_json_array(nested_entry.take())?,
-                false => nested_entry.take(),
-            };
-            let converted_item: T = serde_json::from_value(parsed_item_value)?;
-            flattened_vec.push(converted_item);
-        }
-
-        Ok(Self { flattened_vec })
+        let items = try_parse_array_into_item_from_response_str(str_to_parse)?;
+        Ok(Self { items })
     }
 
     pub fn into_vec(self) -> Vec<T> {
-        self.flattened_vec
+        self.items
     }
 }
 
-fn unwrap_item_in_nested_json_array(mut nested_item: Value) -> Result<Value, ParseError> {
-    let nested_array = nested_item
+fn try_parse_array_into_item_from_response_str<T: de::DeserializeOwned>(
+    json_str: &str,
+) -> Result<Vec<T>, ParseError> {
+    let mut parse_response = serde_json::from_str::<Value>(json_str)?.take();
+    let json_array = parse_response
         .as_array_mut()
-        .ok_or_else(|| generate_none_error("invalid initial json array"))?;
-    match nested_array.last() {
-        Some(_) => Ok(nested_array.pop().unwrap()),
-        None => Ok(json!("")),
+        .ok_or_else(|| generate_none_error("cannot parse initial json string as array"))?;
+
+    let mut item_vec = Vec::new();
+    for json_item in json_array {
+        let converted_item: [T; 1] = serde_json::from_value(json_item.take())?;
+        for item in IntoIter::new(converted_item) {
+            item_vec.push(item);
+        }
```

Comment:
This little bit of code took a while to come up with because of how annoying it is to work with both generics _and_ borrowed `serde_json::Value`s.

Basically by making the type hint be `[T; 1]` we allow the `from_value` call to automatically unpack the `Value::Array([String])` into a `[String]` (nice!).

The following `IntoIter::new` call is just to move the item out of the array (we can't slice because we don't _own_ the data within the slice).